### ISSUE TITLE
Exception filters is not working with fastify v3

### ISSUE THREAD

**Issue by numToStr** [nestjs/nest#5341]
## Current behavior
Exception filter is not working after upgrading to fastify v3. Previously with fastify v2 is working fine. The error is originated from `auth.middleware.ts` which should be catched by `exception.filter.ts`. Also, `res.status` and `res.send` is also undefined inside the exception filter.
 
## Input Code
`server.ts`
```ts
const httpAdapter = app.getHttpAdapter();

app
     // Handling Errors globally overriding nestjs inbuilt ExceptionHandler
     .useGlobalFilters(new AllExceptionsFilter(httpAdapter))
     .setGlobalPrefix("/api");

```

`exception.filter.ts`
```ts
import { FastifyReply, FastifyRequest } from "fastify";
import {
    Catch,
    ArgumentsHost,
    HttpException,
    UnprocessableEntityException,
} from "@nestjs/common";
import { BaseExceptionFilter } from "@nestjs/core";
import { logger } from "../fns/logger";

@Catch()
export class AllExceptionsFilter extends BaseExceptionFilter {
    catch(exception: Error, host: ArgumentsHost): unknown {
        // Goal:
        // I need to catch the UnhandledPromiseRejection or any other which are not nestjs exceptions
        // So, I am checking if the excptions are instance of HttpException,
        // If true, then it means that it is a nestjs Exception [https://docs.nestjs.com/exception-filters]
        // else, I am creating a Exception which I am sending to the client
        const h = host.switchToHttp();
        const res = h.getResponse<FastifyReply>();
        const req = h.getRequest<FastifyRequest>();

        const log = { req: req.raw };

        // All nestjs exceptions are instanceof HttpException
        // So we can easily delegate those exception to nestjs
        if (exception instanceof HttpException) {
            // Error logging
            logger.error(
                Object.assign(log, { res: exception.getResponse() }),
                exception.message
            );

            return super.catch(exception, host);
        }

        // IMPORTANT:
        // If its unknown exception i.e. UnhandledPromiseRejection or FirebaseAuthError

        const r = new UnprocessableEntityException(exception.message);
        const rr = r.getResponse();

        // Error logging
        logger.error(Object.assign(log, { res: rr }), r.message);

        // res.status and res.send in undefined
        return res.status(r.getStatus()).send(rr);
    }
}

```
`auth.middleware.ts`
```ts

@Injectable()
export class AuthMiddleware implements NestMiddleware {
    constructor(
        @Inject("SessionStore") private store: SessionStore<SessionPayload>
    ) {}

    async use(
        req: IMiddlewareRequest,
        _: IMiddlewareReply,
        next: Func
    ): Promise<void> {
        try {
            // Get the token from the header
            // Verify/Validate the token
            // Attached the decoded token to `req.user`
            // You can get the user elsewhere by accessing `req.raw.user`

            const header = req.headers.authorization;

            if (!header) {
                throw new ForbiddenException();
            }

            const [, token] = header.split(" ");

            const decoded = await this.store.get(token);

            req.USER = decoded;

            next();
        } catch {
            throw new UnauthorizedException(
                "Unauthorized Access! Please login."
            );
        }
    }
}

```

## Logs
```logs
server | [1598514052290] ERROR (@cricuru/server/1143 on 63f587120b66): unhandledRejection
server |     TypeError: reply_1.Reply is not a constructor
server |         at FastifyAdapter.reply (/usr/src/app/node_modules/@nestjs/platform-fastify/adapters/fastify-adapter.js:32:15)
server |         at AllExceptionsFilter.catch (/usr/src/app/node_modules/@nestjs/core/exceptions/base-exception-filter.js:26:24)
server |         at AllExceptionsFilter.catch (/usr/src/app/packages/server/src/utils/filters/exception.filter.ts:34:31)
server |         at ExceptionsHandler.invokeCustomFilters (/usr/src/app/node_modules/@nestjs/core/exceptions/exceptions-handler.js:33:26)
server |         at ExceptionsHandler.next (/usr/src/app/node_modules/@nestjs/core/exceptions/exceptions-handler.js:13:18)
server |         at /usr/src/app/node_modules/@nestjs/core/router/router-proxy.js:13:35
server |         at processTicksAndRejections (internal/process/task_queues.js:97:5)
server | [1598514052310] ERROR (@cricuru/server/1143 on 63f587120b66): uncaughtException
server |     TypeError: reply_1.Reply is not a constructor
server |         at FastifyAdapter.reply (/usr/src/app/node_modules/@nestjs/platform-fastify/adapters/fastify-adapter.js:32:15)
server |         at AllExceptionsFilter.catch (/usr/src/app/node_modules/@nestjs/core/exceptions/base-exception-filter.js:26:24)
server |         at AllExceptionsFilter.catch (/usr/src/app/packages/server/src/utils/filters/exception.filter.ts:34:31)
server |         at ExceptionsHandler.invokeCustomFilters (/usr/src/app/node_modules/@nestjs/core/exceptions/exceptions-handler.js:33:26)
server |         at ExceptionsHandler.next (/usr/src/app/node_modules/@nestjs/core/exceptions/exceptions-handler.js:13:18)
server |         at /usr/src/app/node_modules/@nestjs/core/router/router-proxy.js:13:35
server |         at processTicksAndRejections (internal/process/task_queues.js:97:5)
server | /usr/src/app/packages/server/src/bin/www.ts:24
server |         throw err;
server |         ^
server |
server | TypeError: reply_1.Reply is not a constructor
server |     at FastifyAdapter.reply (/usr/src/app/node_modules/@nestjs/platform-fastify/adapters/fastify-adapter.js:32:15)
server |     at AllExceptionsFilter.catch (/usr/src/app/node_modules/@nestjs/core/exceptions/base-exception-filter.js:26:24)
server |     at AllExceptionsFilter.catch (/usr/src/app/packages/server/src/utils/filters/exception.filter.ts:34:31)
server |     at ExceptionsHandler.invokeCustomFilters (/usr/src/app/node_modules/@nestjs/core/exceptions/exceptions-handler.js:33:26)
server |     at ExceptionsHandler.next (/usr/src/app/node_modules/@nestjs/core/exceptions/exceptions-handler.js:13:18)
server |     at /usr/src/app/node_modules/@nestjs/core/router/router-proxy.js:13:35
server |     at processTicksAndRejections (internal/process/task_queues.js:97:5)

```

## Expected behavior
Should work just fine.

## Environment

"@nestjs/common": "^7.4.2",
"@nestjs/core": "^7.4.2",
"@nestjs/platform-fastify": "^7.4.2",
"fastify": "^3.2.1",
 
For Tooling issues:
- Node version: 12.18.2
- Platform: Manjaro Linux

### **Edit: I disabled the custom exception filter, but still the issue is there.**

**Comment by kamilmysliwiec** [comment]
Please provide a minimum reproduction repository.

**Comment by numToStr** [comment]
Hi @kamilmysliwiec, I've created the reproduction in https://github.com/numToStr/nest-exception-filter-bug.

## Bug Reproduction

This is OK

```sh
curl http://localhost:5000 -H 'Authorization:any_random_string'
```

This will crash the server

```sh
curl http://localhost:5000
```

**Comment by numToStr** [comment]
Hi @kamilmysliwiec, Is there any progress on this issue. This issue is forcing me to downgrade to fastify v2.

**Comment by kamilmysliwiec** [comment]
Fixed in 7.4.4

**Comment by numToStr** [comment]
Thanks @kamilmysliwiec
